// @stream-io/logger/dist/cjs/index.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_test_1 = require("node:test");
const node_assert_1 = __importDefault(require("node:assert"));
const index_ts_1 = require("./index.js");
const capture = () => {
    const capturedLogs = {
        trace: [],
        debug: [],
        info: [],
        warn: [],
        error: [],
    };
    const sink = (level, message) => capturedLogs[level].push(message);
    return { sink, capturedLogs };
};
const LOG_REGEX = /^\[([^\]]+)\](?:\(([^)]*)\))?:\s*(.*)$/;
const parseLog = (msg) => {
    const [, scope, tags, message] = msg?.match(LOG_REGEX) ?? [];
    return { scope, tags, message };
};
(0, node_test_1.describe)("logger", () => {
    (0, node_test_1.it)("exposes log level per scope", () => {
        const { sink } = capture();
        const loggerSystem = (0, index_ts_1.createLoggerSystem)();
        loggerSystem.configureLoggers({
            default: { level: "info", sink },
            "coordinator": { level: "debug", sink },
            "network": { level: "error", sink },
        });
        const coordinatorLogger = loggerSystem.getLogger("coordinator");
        const networkLogger = loggerSystem.getLogger("network");
        const apiCallsLogger = loggerSystem.getLogger("api-calls");
        node_assert_1.default.equal(coordinatorLogger.getLogLevel(), "debug");
        node_assert_1.default.equal(networkLogger.getLogLevel(), "error");
        node_assert_1.default.equal(apiCallsLogger.getLogLevel(), "info");
    });
    (0, node_test_1.it)("checks if the message in in correct format", () => {
        const { sink, capturedLogs } = capture();
        const loggerSystem = (0, index_ts_1.createLoggerSystem)();
        loggerSystem.configureLoggers({
            default: { level: "info", sink },
        });
        loggerSystem.getLogger("api-calls", { tags: ["v1", "v2"] }).info("Calling api");
        const msg = capturedLogs.info[0];
        const { scope, tags, message } = parseLog(msg);
        node_assert_1.default.equal(message, "Calling api");
        node_assert_1.default.equal(scope, "api-calls");
        node_assert_1.default.equal(tags, "v1,v2");
    });
    (0, node_test_1.it)("filters below default level (info)", () => {
        const { sink, capturedLogs } = capture();
        const loggerSystem = (0, index_ts_1.createLoggerSystem)();
        loggerSystem.configureLoggers({
            default: { level: "info", sink },
            network: { level: "warn", sink },
            scope: { level: "error", sink },
        });
        loggerSystem.getLogger("scope").info("should not be logged");
        loggerSystem.getLogger("scope").error("should be logged");
        loggerSystem.getLogger("api-call").info("calling api");
        loggerSystem.getLogger("coordinator", { tags: ["v2"] }).info("coordinator test");
        loggerSystem.getLogger("network").warn("network call");
        node_assert_1.default.deepEqual(capturedLogs.error, ["[scope]: should be logged"]);
        node_assert_1.default.deepEqual(capturedLogs.info, [
            "[api-call]: calling api",
            "[coordinator](v2): coordinator test",
        ]);
        node_assert_1.default.deepEqual(capturedLogs.warn, ["[network]: network call"]);
    });
    (0, node_test_1.it)("isolates log levels between different logger systems", () => {
        const { sink } = capture();
        const loggerSystemFeeds = (0, index_ts_1.createLoggerSystem)();
        loggerSystemFeeds.configureLoggers({
            "client": { level: "error", sink },
        });
        const loggerSystemVideo = (0, index_ts_1.createLoggerSystem)();
        loggerSystemVideo.configureLoggers({
            "client": { level: "info", sink },
        });
        const loggerSystemChat = (0, index_ts_1.createLoggerSystem)();
        loggerSystemChat.configureLoggers({
            "client": { level: "warn", sink },
        });
        node_assert_1.default.equal(loggerSystemFeeds.getLogger("client").getLogLevel(), "error");
        node_assert_1.default.equal(loggerSystemVideo.getLogger("client").getLogLevel(), "info");
        node_assert_1.default.equal(loggerSystemChat.getLogger("client").getLogLevel(), "warn");
    });
});
